sn_networking/src/transfers.rs

```diff
@@ -0,0 +1,168 @@
+// under the GPL Licence is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied. Please review the Licences for the specific language governing
+// permissions and limitations relating to use of the SAFE Network Software.
+
+use std::collections::BTreeSet;
+
+use crate::Network;
+
+use sn_dbc::{Dbc, DbcId, DbcSecrets, DbcTransaction, DerivationIndex, SignedSpend};
+use sn_protocol::{
+    error::{Error, Result},
+    storage::{try_deserialize_record, DbcAddress, RecordHeader, RecordKind},
+    NetworkAddress, PrettyPrintRecordKey,
+};
+use sn_transfers::wallet::{LocalWallet, Transfer};
+use tokio::task::JoinSet;
+
+impl Network {
+    /// Gets a spend from the Network.
+    pub async fn get_spend(&self, address: DbcAddress, re_attempt: bool) -> Result<SignedSpend> {
+        let key = NetworkAddress::from_dbc_address(address).to_record_key();
+        let record = self
+            .get_record_from_network(key, None, re_attempt)
+            .await
+            .map_err(|_| Error::SpendNotFound(address))?;
+        debug!(
+            "Got record from the network, {:?}",
+            PrettyPrintRecordKey::from(record.key.clone())
+        );
+        let header =
+            RecordHeader::from_record(&record).map_err(|_| Error::SpendNotFound(address))?;
+
+        if let RecordKind::DbcSpend = header.kind {
+            match try_deserialize_record::<Vec<SignedSpend>>(&record)
+                .map_err(|_| Error::SpendNotFound(address))?
+                .as_slice()
+            {
+                [one, two, ..] => {
+                    error!("Found double spend for {address:?}");
+                    Err(Error::DoubleSpendAttempt(
+                        Box::new(one.to_owned()),
+                        Box::new(two.to_owned()),
+                    ))
+                }
+                [one] => {
+                    trace!("Spend get for address: {address:?} successful");
+                    Ok(one.clone())
+                }
+                _ => {
+                    trace!("Found no spend for {address:?}");
+                    Err(Error::SpendNotFound(address))
+                }
+            }
+        } else {
+            error!("RecordKind mismatch while trying to retrieve a Vec<SignedSpend>");
+            Err(Error::RecordKindMismatch(RecordKind::DbcSpend))
+        }
+    }
+
+    /// Verify Transfer and rebuild spendable currency from it
+    /// Returns an `Error::FailedToDecypherTransfer` if the transfer cannot be decyphered
+    /// (This means the transfer is not for us as it was not encrypted to our key)
+    /// Returns an `Error::InvalidTransfer` if the transfer is not valid
+    /// Else returns a list of DBCs that can be deposited to our wallet and spent
+    pub async fn verify_and_unpack_transfer(
+        &self,
+        transfer: Transfer,
+        wallet: &LocalWallet,
+    ) -> Result<Vec<Dbc>> {
```

Comment:
This function to receive a `Transfer` and turn it back into spendable DBCs. **Needs Network connection**.